File: src/main/java/de/lulonaut/Bot/commands/config/ConfigStateMachine.java
package de.lulonaut.Bot.commands.config;

import de.lulonaut.Bot.utils.Cache;
import de.lulonaut.Bot.utils.Conf;
import de.lulonaut.Bot.utils.Database;
import de.lulonaut.Bot.utils.GetGuild;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.MessageChannel;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;

import java.util.HashMap;
import java.util.Map;

public class ConfigStateMachine extends ListenerAdapter {
    private final long channelID, userID;

    private final Map<String, String> options = new HashMap<>();
    public boolean done;
    private GuildMessageReceivedEvent initEvent;
    private int currentStep;
    private boolean inProgress;
    /*
    Steps:
    1: Verified role
    2: message counting system toggle
    3: guild member role toggle
    4: if 3: guild name (validated with MC name)
     */

    public ConfigStateMachine(MessageChannel channel, GuildMessageReceivedEvent event, Member member) {
        System.out.println("starting");
//        this.guildID = guild.getIdLong();
        this.channelID = channel.getIdLong();
        this.userID = member.getIdLong();
        this.currentStep = 1;
        this.done = false;
        this.initEvent = event;
        //default options
        options.put("role", "Hypixel Verified");
        options.put("counting", "false");
        options.put("guildMemberRole", "false");
        options.put("guildID", "");
    }

    private String showCurrentConfig() {
        StringBuilder sb = new StringBuilder();
        String count;
        if (options.get("counting").equals("true")) {
            count = "enabled";
        } else {
            count = "disabled";
        }
        String gRole;
        boolean gmRole;
        if (options.get("guildMemberRole").equals("true")) {
            gRole = "enabled";
            gmRole = true;
        } else {
            gRole = "disabled";
            gmRole = false;
        }
        sb.append("Verified Role: `").append(options.get("role")).append("`\n");
        sb.append("Message counting system: `").append(count).append("`\n");
        sb.append("Guild member role: `").append(gRole).append("`\n");
        if (gmRole) {
            sb.append("Guild Name: `").append(options.get("guildID")).append("`");
        }

        return sb.toString();
    }

    @Override
    public void onMessageReceived(MessageReceivedEvent event) {
        System.out.println("event...");
        if (event.getAuthor().isBot() || event.getChannel().getIdLong() != channelID || event.getAuthor().getIdLong() != userID) {
            System.out.println("return1");
            return;
        }

        if ((event.getMessage().getContentRaw().equalsIgnoreCase("cancel") && !inProgress)) {
            event.getChannel().sendMessage("setup cancelled.").queue();
            event.getJDA().removeEventListener(this);
            System.out.println("return2");
            return;
        }
        if ((event.getMessage().getContentRaw().equalsIgnoreCase("continue") && !inProgress)) {
            System.out.println("continue");
            currentStep++;
        }
        if ((event.getMessage().getContentRaw().equalsIgnoreCase("exit") && !inProgress)) {
            System.out.println(options);
            Database.saveConfig(event.getGuild().getId(), options);
            //TODO: Write changes to Database (new method for saving and getting current config)
            event.getChannel().sendMessage("Config saved to Database.\nCurrent config:" +
                    showCurrentConfig()).queue();
            event.getJDA().removeEventListener(this);
            System.out.println("return3");
            return;
        }
        System.out.println("switch with " + currentStep);
        switch (currentStep) {
            case 1:
                //verify role
                if (inProgress) {
                    options.put("role", event.getMessage().getContentRaw());
                    event.getChannel().sendMessage("OK: Verify Role is now set to: `" + event.getMessage().getContentRaw() + "` Please make sure this role exists before proceeding\nType `continue` to continue with the setup, `exit` to save the config and exit or `cancel` to exit without saving.").queue();
                    inProgress = false;
                } else {
                    event.getChannel().sendMessage("Welcome to the config setup\nPlease specify a Role users get when verifying successfully. This is currently set to `" + options.get("role") + "`").queue();
                    inProgress = true;
                }
                break;
            case 2:
                //message counting
                if (inProgress) {
                    if (event.getMessage().getContentRaw().equalsIgnoreCase("enabled")) {
                        options.put("counting", "true");
                        event.getChannel().sendMessage("OK: Message counting system is now enabled.\n" +
                                "With it come some commands:\n" +
                                Conf.PREFIX + "`leaderboard`: shows the current leaderboard of users with the most messages\n" +
                                Conf.PREFIX + "`check <user>`: shows how many messages you or a given user currently have\n" +
                                Conf.PREFIX + "`deleteMessages`: this command requires the \"Manage Server\" Permission and clears the current message counter for everyone in the server. (further confirmation is required)\n" +
                                "\n" +
                                "Please type `continue` to continue with the setup, `exit` to save the config and exit or `cancel` to exit without saving.").queue();
                        inProgress = false;
                    } else if (event.getMessage().getContentRaw().equalsIgnoreCase("disabled")) {
                        options.put("counting", "false");
                        event.getChannel().sendMessage("OK: Message counting system is now disabled.\n" +
                                "Please type `continue` to continue with the setup, `exit` to save the config and exit or `cancel` to exit without saving.").queue();
                        inProgress = false;
                    } else {
                        event.getChannel().sendMessage("Invalid input, please type `enabled` or `disabled`.").queue();
                    }
                } else {
                    event.getChannel().sendMessage("The message counting system tracks every message from every user and allows for a leaderboard, individual checking for a user and resetting of the counter.\n" +
                            "type one of the following: `enabled` to enable it or `disabled` to disable it.").queue();
                    inProgress = true;
                }
                break;
            case 3:
                //guild member role toggle
                if (inProgress) {
                    if (event.getMessage().getContentRaw().equalsIgnoreCase("enabled")) {
                        options.put("guildMemberRole", "true");
                        event.getChannel().sendMessage("OK: guild member role enabled. In the next step you will have to set a guild name. If you skip this step the feature will not work.\n" +
                                "Please type `continue` to continue with the setup, `exit` to save the config and exit or `cancel` to exit without saving.").queue();
                        inProgress = false;
                        return;
                    } else if (event.getMessage().getContentRaw().equalsIgnoreCase("disabled")) {
                        options.put("guildMemberRole", "false");
                        event.getChannel().sendMessage("OK: guild member role disabled.\n" +
                                "Please type `continue` to continue with the setup, `exit` to save the config and exit or `cancel` to exit without saving.").queue();
                        currentStep++;
                        inProgress = false;
                        return;
                    } else {
                        event.getChannel().sendMessage("Invalid input, please type `enabled` or `disabled`.").queue();
                    }
                } else {
                    event.getChannel().sendMessage("Setting up a guild role allows you to skip giving each guild member a role by hand if you run a Discord for a guild. If you don't this feature is pretty useless.\n" +
                            "Type one of the following: `enabled` to enable it or `disabled` to disable it.").queue();
                    inProgress = true;
                }
                break;
            case 4:
                //getting guild name
                if (inProgress) {
                    String guildName;
                    try {
                        guildName = GetGuild.getGuild(event.getMessage().getContentRaw());
                    } catch (Exception e) {
                        event.getChannel().sendMessage("An error occurred while validating the guild, defaulting to no guild role for now. Please try again later if you think this is a mistake\n" +
                                "Please type `continue` to continue with the setup, `exit` to save the config and exit or `cancel` to exit without saving.").queue();
                        options.put("guildMemberRole", "disabled");
                        inProgress = false;
                        return;
                    }
                    if (guildName.equals("")) {
                        event.getChannel().sendMessage("Either this user is not in a guild or an error occurred while checking this. Please try again later if you believe this is a mistake. Defaulting to no guild role for now.\n" +
                                "Please type `continue` to continue with the setup, `exit` to save the config and exit or `cancel` to exit without saving.").queue();
                        options.put("guildMemberRole", "disabled");
                    } else {
                        options.put("guildID", guildName);
                        event.getChannel().sendMessage("OK: Guild Name is now set to: `" + guildName + "`\n" +
                                "Please type `continue` to continue with the setup, `exit` to save the config and exit or `cancel` to exit without saving.").queue();
                    }
                    inProgress = false;
                    return;

                } else {
                    event.getChannel().sendMessage("Please type your Minecraft username to set your guild.").queue();
                    inProgress = true;
                }
                break;
            case 5:
                //final message
                System.out.println(options);
                Database.saveConfig(event.getGuild().getId(), options);
                Cache.refreshOrAddCache(event.getGuild().getId());
                EmbedBuilder eb = new EmbedBuilder();
                eb.setTitle("Setup done");
                eb.setDescription("Setup finished.\n" +
                        "These are the current config options for " + event.getGuild().getName() + ":\n" +
                        showCurrentConfig());
                event.getChannel().sendMessage(eb.build()).queue();
                done = true;
                event.getJDA().removeEventListener(this);
        }
    }
}

File: src/main/kotlin/de/lulonaut/bot/commands/config/ConfigStateMachine.kt
package de.lulonaut.bot.commands.config

import de.lulonaut.bot.cache.DatabaseCache
import de.lulonaut.bot.utils.*
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.MessageChannel
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter

class ConfigStateMachine(channel: MessageChannel, event: GuildMessageReceivedEvent, member: Member?) :
    ListenerAdapter() {
    private val channelID: Long = channel.idLong
    private val userID: Long = member!!.idLong
    private var options: MutableMap<String?, String?>? = HashMap()
    var done: Boolean
    private val initEvent: GuildMessageReceivedEvent
    private var currentStep: Int = 1
    private var inProgress = false


    private fun showCurrentConfig(): String {
        val sb = StringBuilder()

        val gmRole: Boolean = this.options?.get("guildMemberRole") == "true"
        sb.append("Verified Role: `").append(this.options?.get("role")).append("`\n")
        sb.append("Rank roles: `").append(this.options?.get("rankRoles")).append("`\n")
        sb.append("Message counting system: `").append(this.options?.get("counting")).append("`\n")
        sb.append("Guild member role: `").append(this.options?.get("guildMemberRole")).append("`\n")
        if (gmRole) {
            sb.append("Guild Name: `").append(this.options?.get("guildID")).append("`\n")
            sb.append("Guild member role: `").append(this.options?.get("guildRoleName")).append("`\n")
        }
        sb.append("Command prefix: `").append(this.options?.get("prefix")).append("`")
        return sb.toString()
    }

    override fun onMessageReceived(event: MessageReceivedEvent) {
        if (event.author.isBot || event.channel.idLong != channelID || event.author.idLong != userID) {
            return
        }
        if (event.message.contentRaw.equals("continue", ignoreCase = true) && !inProgress) {
            currentStep++
        }
        if (event.message.contentRaw.equals("exit", ignoreCase = true) && !inProgress) {
            Database.saveConfig(event.guild.id, this.options)
            val eb = EmbedBuilder()
            eb.setTitle("Setup closed")
            eb.setDescription(
                "Config saved to Database.\n" +
                        "Current config for ${event.guild.name}:\n" +
                        showCurrentConfig()
            )
            event.channel.sendMessage(eb.build()).queue()
            event.jda.removeEventListener(this)
            this.options = null
            return
        }
        when (currentStep) {
            1 ->                 //verify role
                if (inProgress) {
                    this.options?.set("role", event.message.contentRaw)
                    event.channel.sendMessage(
                        """
    OK: Verify Role is now set to: `${event.message.contentRaw}` Please make sure this role exists before proceeding
    Type `continue` to continue with the setup or  `exit` to save the config.
    """.trimIndent()
                    ).queue()
                    inProgress = false
                } else {
                    event.channel.sendMessage(
                        """
    Welcome to the config setup
    Please specify a Role users get when verifying successfully. This is currently set to `${this.options?.get("role")}`
    """.trimIndent()
                    ).queue()
                    inProgress = true
                }
            2 ->                 //message counting
                if (inProgress) {
                    when {
                        event.message.contentRaw.equals("enabled", ignoreCase = true) -> {
                            this.options?.set("counting", "true")
                            event.channel.sendMessage(
                                "OK: Message counting system is now enabled. " +
                                        "With it come some commands:\n" +
                                        "${Conf.PREFIX}`leaderboard`: shows the current leaderboard of users with the most messages\n" +
                                        "${Conf.PREFIX}`check <user>`: shows how many messages you or a given user currently have\n" +
                                        "${Conf.PREFIX}`deleteMessages`: this command requires the \"Manage Server\" Permission and clears the current message counter for everyone in the server. (further confirmation is required)\n" +

                                        "Type `continue` to continue with the setup or  `exit` to save the config"
                            ).queue()
                            inProgress = false
                        }
                        event.message.contentRaw.equals("disabled", ignoreCase = true) -> {
                            this.options?.set("counting", "false")

                            event.channel.sendMessage(
                                "OK: message counting system is now disabled\n" +
                                        "Type `continue` to continue with the setup or  `exit` to save the config"
                            ).queue()
                            inProgress = false
                        }
                        else -> {
                            event.channel.sendMessage("Invalid input, please type `enabled` or `disabled`.").queue()
                        }
                    }
                } else {
                    event.channel.sendMessage(
                        """
    The message counting system tracks every message from every user and allows for a leaderboard, individual checking for a user and resetting of the counter.
    type one of the following: `enabled` to enable it or `disabled` to disable it.
    """.trimIndent()
                    ).queue()
                    inProgress = true
                }
            3 ->                 //guild member role toggle
                if (inProgress) {
                    when {
                        event.message.contentRaw.equals("enabled", ignoreCase = true) -> {
                            this.options?.set("guildMemberRole", "true")
                            event.channel.sendMessage(
                                """
                    OK: guild member role enabled. In the next step you will have to set a guild name. If you skip this step the feature will not work.
                    Type `continue` to continue with the setup or  `exit` to save the config
                    """.trimIndent()
                            ).queue()
                            inProgress = false
                            return
                        }
                        event.message.contentRaw.equals("disabled", ignoreCase = true) -> {
                            this.options?.set("guildMemberRole", "false")
                            event.channel.sendMessage(
                                """
                    OK: guild member role disabled.
                    Type `continue` to continue with the setup or  `exit` to save the config
                    """.trimIndent()
                            ).queue()
                            currentStep += 2
                            inProgress = false
                            return
                        }
                        else -> {
                            event.channel.sendMessage("Invalid input, please type `enabled` or `disabled`.").queue()
                        }
                    }
                } else {
                    event.channel.sendMessage(
                        """
    Setting up a guild role allows you to skip giving each guild member a role by hand if you run a Discord for a guild. If you don't this feature is pretty useless.
    Type one of the following: `enabled` to enable it or `disabled` to disable it.
    """.trimIndent()
                    ).queue()
                    inProgress = true
                }
            4 ->                 //getting guild name
                if (inProgress) {
                    val guildName: String?
                    try {
                        guildName = API.getGuild(event.message.contentRaw)
                    } catch (e: Exception) {
                        event.channel.sendMessage(
                            """
    An error occurred while validating the guild, defaulting to no guild role for now. Please try again later if you think this is a mistake
    Type `continue` to continue with the setup or  `exit` to save the config
    """.trimIndent()
                        ).queue()
                        this.options?.set("guildMemberRole", "disabled")
                        inProgress = false
                        return
                    }
                    if (guildName == "") {
                        event.channel.sendMessage(
                            """
    Either this user is not in a guild or an error occurred while checking this. Please try again later if you believe this is a mistake. Defaulting to no guild role for now.
    Type `continue` to continue with the setup or  `exit` to save the config
    """.trimIndent()
                        ).queue()
                        this.options?.set("guildMemberRole", "disabled")
                        currentStep++
                    } else {
                        this.options?.set("guildID", guildName)
                        event.channel.sendMessage(
                            """
    OK: Guild Name is now set to: `$guildName` Now you need to set a role players get when they are in the guild.
    Type `continue` to continue with the setup or  `exit` to save the config
    """.trimIndent()
                        ).queue()
                    }
                    inProgress = false
                    return
                } else {
                    event.channel.sendMessage("Please type your Minecraft username to set your guild.").queue()
                    inProgress = true
                }
            5 -> {
                if (inProgress) {
                    this.options?.set("guildRoleName", event.message.contentRaw)
                    event.channel.sendMessage(
                        "OK: Guild Member role is now set to: `${event.message.contentRaw}`\n" +
                                "Type `continue` to continue with the setup or  `exit` to save the config"
                    ).queue()
                    inProgress = false
                } else {
                    event.channel.sendMessage("Please provide a role that is assigned to players that are in your guild.")
                        .queue()
                    inProgress = true
                }
            }
            6 -> {      //rank role toggle
                val message: String = event.message.contentRaw
                if (inProgress) {
                    if (message != "enabled" && message != "disabled") {
                        event.channel.sendMessage("Invalid input : `${event.message.contentRaw}`. Choose either `enabled` or `disabled`")
                            .queue()
                        return
                    } else {
                        if (message == "enabled") {
                            this.options?.set("rankRoles", "true")
                        } else {
                            this.options?.set("rankRoles", "false")
                        }
                    }
                    event.channel.sendMessage(
                        "OK: Rank roles are now ${message}\n" + "Type `continue` to continue with the setup or  `exit` to save the config"
                    ).queue()
                    inProgress = false
                } else {
                    event.channel.sendMessage(
                        "The bot can check the current rank of a user and give them the role for this rank." +
                                "This requires the following roles to be created:\n" +
                                "`VIP`\n" +
                                "`VIP+`\n" +
                                "`MVP`\n" +
                                "`MVP+`\n" +
                                "`MVP++`\n\n" +
                                "If you want this enabled type `enabled` otherwise `disabled`."
                    ).queue()
                    inProgress = true
                }
            }
            7 -> {      //prefix
                inProgress = if (inProgress) {
                    this.options?.set("prefix", event.message.contentRaw)
                    event.channel.sendMessage(
                        "OK: Prefix is now set to: `${event.message.contentRaw}`." +
                                "Type `continue` to continue with the setup or  `exit` to save the config"
                    ).queue()
                    false
                } else {
                    val currentPrefix = this.options?.get("prefix")
                    event.channel.sendMessage(
                        "Now please set a prefix to be used." +
                                "The current Prefix is `$currentPrefix`\n"
                    ).queue()
                    true
                }
            }

            8 -> {
                //final message
                Database.saveConfig(event.guild.id, this.options)
                DatabaseCache.refreshOrAddCache(event.guild.id)
                val eb = EmbedBuilder()
                eb.setTitle("Setup finished")
                eb.setDescription(
                    """
    These are the current config this.options for ${event.guild.name}:
    ${showCurrentConfig()}
    """.trimIndent()
                )
                event.channel.sendMessage(eb.build()).queue()
                done = true
                this.options = null
                event.jda.removeEventListener(this)
            }
            else -> {
                event.channel.sendMessage("Invalid input.")
            }
        }
    }

    /*
    Steps:
    1: Verified role
    2: message counting system toggle
    3: guild member role toggle
    4: if 3: guild name (validated with MC name)
    5: rank roles toggle
    6: command prefix
     */
    init {
        //        this.guildID = guild.getIdLong();
        done = false
        initEvent = event
        //default this.options
        this.options = DatabaseCache.getConfig(event.guild.id) as MutableMap<String?, String?>?
        if (!this.options?.contains("prefix")!!) {
            this.options?.set("role", "Hypixel Verified")
            this.options?.set("counting", "false")
            this.options?.set("guildMemberRole", "false")
            this.options?.set("guildID", "")
            this.options?.set("prefix", Conf.PREFIX)
            this.options?.set("rankRoles", "false")
        }

    }
}